### ISSUE TITLE
ISO-BMFF encapsulation should define a box inside the AudioSampleEntry for the configOBUs

### ISSUE THREAD

**Issue by jamrial** [AOMediaCodec/iamf#792]
Currently, the ISO-BMFF IAMF encapsulation is defined as follows:

```
class IASampleEntry extends AudioSampleEntry('iamf') {
    unsigned int (8) configOBUs[];
}
```

Said sample entry being in the stsd box. The problem here is that AudioSampleEntry can have other boxes inside it, like btrt (see section 8.5.2.2 of ISO 14496-12), and if said box is present, there's nothing delimiting where configOBUs ends. A parser may keep reading sample entry bytes after the last valid OBU, and will in turn try to parse a btrt box as if it was an OBU, with unpredictable results.

The solution is simple, albeit a problem considering the spec was finalized: Define a box inside the Sample Entry for the configOBUs, like it was done for example for AV1 in ISOBMFF (https://aomediacodec.github.io/av1-isobmff/#av1codecconfigurationbox-definition)

In short, something like:

```
class IASampleEntry extends AudioSampleEntry('iamf') {
    IAConfigurationBox config;
}

class IAConfigurationBox extends Box('iacb') {
    unsigned int (8) configOBUs[];
}
```

If changing how the existing Sample Entry is defined is not an option, then perhaps defining a new one and releasing a 1.0.1 version of the spec, stating it must be used when optional boxes within the AudioSampleEntry are present.

**Comment by sunghee-hwang** [comment]
I think what IASampleEntry specifies is to just add configOBUs to a given AudioSampleEntry whatever it is. 

AudioSampleEntry is defined according to the rule specified in section 12.2.3.2 of ISO 14496-12. 
When "btrt" or any optional box is present, it is based on the rule in section 12.2.3.2 of ISO 14496-12, and so the box is followed by configOBUs[]. 

Based on this, I don't think that the current spec has a problem to parse.

Please let me know if I am wrong.

**Comment by jamrial** [comment]
As far as i can see, these boxes can go anywhere within a Sample Entry after samplerate (Last mandatory field), be it before or after extension defined entries like IASampleEntry. Same case with colr, clap, pasp, etc, in VisualSampleEntry, where for h264/hevc tracks they can go after the avcC/hvcC box (Which is what writers like GPAC do).

**Comment by jamrial** [comment]
To exemplify, how can a parser know where the optional boxes stop and the ConfigOBUs start if the latter were to be last? There could be a btrt box, or a srat box, or both. By looking for boxes (size field + box name field) it can parse them just fine, but if ConfigOBUs suddenly start with no warning after samplerate or any of those optional boxes, it would have no choice other than doing heuristics and see if the bytes look like OBUs, which has a high chance of hitting false positives.

Everything after samplerate needs to be its own box.

**Comment by sunghee-hwang** [comment]
From the ISOBMFF point of view, does it mean that the optional boxes within AudioSampleEntry can be placed in random order regardless of the order listed within section 12.2.3.2 Syntax of ISO 14496-12?
For example, even though ChannelLayout() is listed before DownMixInstructions() within the class AudioSampleEntry() of section 12.2.3.2, does it mean that ChannelLayout() may be present after DownMixInstructions()?  

Regardless of the above question, the possible value of the first byte of each config OBU (IA Sequence Header, Codec Config, Audio Element, or Mix Presentation) is 0x0x, 0x1x, or 0xFx. So, it can be distinguished from any coding name of ISOBMFF.

FYI, an old version of the current IAMF spec had your proposed structure. However, it has been updated to the current structure.

**Comment by jamrial** [comment]
> For example, even though ChannelLayout() is listed before DownMixInstructions() within the class AudioSampleEntry() of section 12.2.3.2, does it mean that ChannelLayout() may be present after DownMixInstructions()?

Those two specific boxes may need to be in that order if they are present (as they are optional), but anything else is "Further boxes as needed", which includes extensions like IASampleEntry, and hints that it needs to define a box.
This is for that matter the only spec i can think about where a Sample Entry extension does not define a box for its own codec specific fields.

> Regardless of the above question, the possible value of the first byte of each config OBU (IA Sequence Header, Codec Config, Audio Element, or Mix Presentation) is 0x0x, 0x1x, or 0xFx. So, it can be distinguished from any coding name of ISOBMFF.

First four bytes for a box are the size. This value can be anything, including something a parser's heuristics can interpret as an OBU header.

**Comment by sunghee-hwang** [comment]
> First four bytes for a box are the size. This value can be anything, including something a parser's heuristics can interpret as an OBU header.
You are right. I misunderstood it.

**Comment by sunghee-hwang** [comment]
When we add configOBUs as a box (e.g., 'iacb'), would you let me know why you think that 'iacb' can come before 'btrt' box?
My understanding is that:
- 'iacb' is added according to not the syntax of AudioSampleEntry but the syntax of IASampleEntry. So, it could not be a box for "further boxes as needed".
- 'btrt' is added according to the syntax of AudoSampleEntry. So, it is a box for "further boxes as needed".

Do you think that 'iacb' can be regarded as a box for "further boxes as needed"?

**Comment by jamrial** [comment]
> Do you think that 'iacb' can be regarded as a box for "further boxes as needed"?

It can. It's what most real world implementations of the spec do with other Sample Entry definitions, e.g. writers like GPAC. pasp, colr and btrt boxes are usually placed after avcC/hvcC, etc.

And again, even if there were no optional boxes in a given Sample Entry in a file, a parser can't know that. It can only be sure of one thing: As long as Sample Entry bytes remain to be read, the first byte after samplerate can (or rather, has to) be part of a size.
configOBUs as is currently defined goes against that. Is this 0xF8 part of a size for an optional box, or the start of a Sequence Header in configOBUs? A parser shouldn't have to make assumptions or guesses, it should only parse a defined struct. A box containing the configOBUs ensures that.

**Comment by sunghee-hwang** [comment]
I understand the situation and agree we must have it as a box. And, also, we need to make the location of the extended box within IASampleEntry clear.

It seems to me that AudioSampleEntry does not allow the extension of anything that is not the box structure, especially outside the ISOBMFF specification. And, it insists that any extension should be a box for "further boxes as needed". 
I found the sentence in ISOBMFF spec saying "For maximum compatibility, the SamplingRateBox, ChannelLayoutBox
and any DownMix and DRC boxes should follow, not precede, any boxes defined in or required by
derived specifications."

In that sense, IMHO, the syntax structure, "class xxxSampleEntry extends yyySampleEntry () {}", has been misused to say that an extended box is regarded as a box for "further boxes as needed".

What do you think?

**Comment by sunghee-hwang** [comment]
Thanks for your opinion.

Anyway, I am trying to find a way to solve this issue with no updating of the IAMF spec or minimizing the updates.

I found the following paragraph in section 8.5.2.1
```
All SampleEntry boxes may contain “extra boxes” not explicitly defined in the box syntax of this or
derived specifications. When present, such boxes shall follow all defined fields and should follow any
defined contained boxes. Decoders shall presume a sample entry box could contain extra boxes and
shall continue parsing as though they are present until the containing box length is exhausted.
```

As the current IASampleEntry has configOBUs not as a box but as fields, the paragraph implies that configOBUs come immediately after the `samplerate` field and are followed by boxes for "further boxes as needed". 

FYI, see #403.

@cconcolato, @tdaede 
can you check this?

**Comment by sunghee-hwang** [comment]
You are right. Logically, I agree that a parser can't know where configOBUs end.
The case in which a parser misunderstands as if it were an OBU is that the first 5 bits of a box size are 0b00010 (i.e., the OBU type for Mix Presentation OBU). This means that the box size is bigger or equal to 64MB (64 x 1024 x 1024).

In this sense, one solution could be to restrict the size of any optional box to less than 64MB when an IAMF ISOBMFF file is generated.
In a real situation, its size would be much less small.

What do you think?

**Comment by jamrial** [comment]
No, as someone that's writing an implementation for this spec i don't agree about any kind of heuristics or extra checks outside of the norm being required for parsers in the spec. I already expressed above that a parser should only handle a defined struct and not try to make any assumptions. Reading a byte to check if they may be the start of a size or an OBU header should not be needed in this format.

I know i'm not the one who will ultimately make the decision of how to move forward with this issue, but i insist, the spec is fresh, real world usage still doesn't exist, new conformance samples can be easily made, so a swiftly released update introducing a new Sample Entry should be painless and very welcome to implementations.

**Comment by sunghee-hwang** [comment]
Let me ask one thing regarding the definition of the box size in the ISOBMFF spec.
For a given type 'abcd', should parsers assume that the box size points to the end of the syntaxes that parsers know about? 
In other words, does the spec disallow to add any extra bytes to the end of the box with the box type = 'abcd'?

**Comment by jamrial** [comment]
I have no authority to say what it disallows, but it's evident to me that the size pointing to the end of the box and all syntax elements being defined means no extra undefined bytes are meant to be present at the end. A box either has defined fields, child boxes, or both, allowing parsers to reach the end of the box by reading them.
If the box allows optional child boxes, there can't be undefined bytes at the end because a parser will expect them to be a box.

**Comment by sunghee-hwang** [comment]
Thank you for your opinions.

**Comment by sunghee-hwang** [comment]
I had some discussion with @jwcullen.
As a result, I'd like to share two options to resolve this issue.

- Option1: follow Jamrial's suggestion. (the same level as AV1 and the old version of IAMF took this)
- Option2: add a signal to indicate the end of configOBUs directly. (more robust than AV1)

```
class IASampleEntry extends AudioSampleEntry('iamf') {
    IAConfigurationBox config;
}

class IAConfigurationBox extends Box('iacb') {
    unsigned int (8) num_OBUs; // or unsigned int (32) configOBUs_size; 
    unsigned int (8) configOBUs[];
}
```

**Comment by jamrial** [comment]
configOBUs_size is superfluous when the "iacb" box already has a size field that will essentially be the same value.

No strong opinion about num_OBUs. I guess it can work as an extra layer to ensure integrity (parsers can warn or abort if they find less configOBUs than the signaled amount). It will not really aid in parsing aside from that.

**Comment by sunghee-hwang** [comment]
After taking option 1 for V1 spec, will it be non-compliant with the ISOBMFF rule if we add a child box of 'iacb' in V2 spec?

**Comment by jamrial** [comment]
You're right. If you plan to extend 'iacb' then you need to signal where configOBUs end. I however don't think adding child boxes to 'iacb' would be necessary or even a good idea. Optional boxes can be in the AudioSampleEntry alongside 'iacb'.

Further fields can be added in the future to IAConfigurationBox and parsers can ignore any such data if they don't recognize it (As in, they adhere to the currently defined syntax) by skipping all bytes in the box past configOBUs. This should be stated.
A version field can also be added if breaking changes were to occur (So parsers don't attempt to read the box at all if they don't recognize the version number).

```
class IAConfigurationBox extends Box('iacb') {
    unsigned int (8) configurationVersion = 1;
    leb128() configOBUs_size; // leb128() should save a few bytes compared to unsigned int (32)
    unsigned int (8) configOBUs[];
}
```

**Comment by tdaede** [comment]
We discussed this and decided that we would be OK with making a breaking change to the box in an update, rather than having to support both the original and fixed box.

**Comment by jwcullen** [comment]
I'm considering taking the simple option here. 

```
class IASampleEntry extends AudioSampleEntry('iamf') {
    IAConfigurationBox config;
}

class IAConfigurationBox extends Box('iacb') {
    unsigned int (8) configOBUs[];
}
```

Rationale:
  - Against `configOBUs_size` 
     - `configOBUs_size`  is redundant because the size can already be inferred from the `iacb` box size.
  - Against `num_OBUs`
    - `num_OBUs` does not seem that valuable. We already have a little bit of redundancy; parsing the cumulative iamf-`obu_size`s already should agree with the `iacb`-box size.
  - Against `configurationVersion`
    - Most extensions we may add in the future probably would live in an OBU anyway - naturally fitting into `iacb`.
    - If we need an extension that can't live in an OBU then jamrial's point is valid: "Optional boxes can be in the AudioSampleEntry alongside 'iacb'."
    - I think these types of extensions are prone to various third-party implementation having bugs/quirks/etc anyways.

**Comment by jwcullen** [comment]
Thanks, jamrial, for all your feedback and input. I could imagine some isobmff specific config updates we might need in the future. And definitely if we use `configurationVersion` we are best off having `configOBUs_size`. 

Since we decided it is OK to make a breaking change - I think your latest suggestion is fine. I will make a PR based on it:
```
class IASampleEntry extends AudioSampleEntry('iamf') {
    IAConfigurationBox config;
}

class IAConfigurationBox extends Box('iacb') {
  IAConfigurationRecord IAMFConfig;
}

aligned(8) class IAConfigurationRecord {
    unsigned int (8) configurationVersion = 1;
    leb128() configOBUs_size; // leb128() should save a few bytes compared to unsigned int (32)
    unsigned int (8) configOBUs[];
}
```

Since this is a breaking change; the working group may need to have one final discussion, update the reference decoder, and test bitstreams before officially approving and merging the PR. I don't expect it to officially be merged in until January.
